homeassistant/components/prometheus/sensor.py

```diff
@@ -0,0 +1,115 @@
+"""Prometheus Sensor component."""
+from datetime import timedelta
+import logging
+from typing import Union
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA as SENSOR_PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_UNIT_OF_MEASUREMENT,
+    CONF_URL,
+    STATE_UNKNOWN,
+)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+DEFAULT_URL = "http://localhost:9090"
+CONF_QUERIES = "queries"
+CONF_EXPR = "expr"
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+_QUERY_SCHEMA = vol.Schema(
+    {
+        vol.Required(
+            CONF_NAME,
+        ): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+        vol.Required(CONF_EXPR): cv.string,
+    }
+)
+
+PLATFORM_SCHEMA = SENSOR_PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_URL, default=DEFAULT_URL): cv.string,
+        vol.Required(CONF_QUERIES): [_QUERY_SCHEMA],
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+    session = async_get_clientsession(hass)
+
+    prometheus = Prometheus(config.get(CONF_URL), session)
+
+    sensors = []
+    for query in config.get("queries", dict()):
+        sensors.append(PrometheusSensor(prometheus, query))
+
+    async_add_entities(sensors, update_before_add=True)
+
+
+class Prometheus:
+    """Wrapper for Prometheus API Requests."""
+
+    def __init__(self, url: str, session) -> None:
+        """Initialize the Prometheus API wrapper."""
+        self._session = session
+        self._url = f"{url}/api/v1/query"
+
+    async def query(self, expr: str) -> Union[str, float]:
+        """Query expression response."""
+        response = await self._session.get(self._url, params={"query": expr})
+        if response.status != 200:
+            _LOGGER.error(
+                "Unexpected HTTP status code %s for expression '%s'",
+                response.status,
+                expr,
+            )
+            return STATE_UNKNOWN
+
+        try:
+            result = (await response.json())["data"]["result"]
+        except (ValueError, KeyError) as error:
+            _LOGGER.error("Invalid query response: %s", error)
+            return STATE_UNKNOWN
+
+        return float(result[0]["value"][1])
```

Comment:
I would advise against ignoring all results other than the first one. This could surprise users, and also I don't think Prometheus really guarantees to return results in the same order.

Two possible solutions here: either return an error if the response contains multiple results (pretty simple) or use all of them (for example, by setting a separate state attribute for each result).

Reply:
There is two unrelated things to consider:

- I think from a series I think I can only consider the last value in a sensor entity, or can I backfill it?

- Expressions need to map to exactly one series or else its value is ambiguous.

Reply:
1. `/api/v1/query` is an instant query API that I believe will return the latest point for each metric. If you wanted to get a time series, you would need to use a range query API (`/api/v1/query_range`), but I agree that it does not seem like that can be useful for a Home Assistant sensor (history in HA seems to be provided by the separate [history](https://www.home-assistant.io/integrations/history/) integration that does not seem to have a backfill API).
2. Yes, that's exactly the issue that I wanted to point out here. I would suggest to return an error unless `result` contains exactly one element.

Reply:
Logging an error and returning `STATE_PROBLEM` if zero or multiple metrics are returned.